Security Config 설정 및 JWT 인증/인가 방식 추가

jinseong205 opened the pull request:
스프링 시큐리티와 JWT 인증/인가 방식 추가하였습니다.
PR을 나눠서 올리기가 애매해서 한번에 올립니다.
다른 방법이나 수정할부분 있으면 코멘트 남겨주시면 감사하겠습니다.

- UserDetails 구현체 PrincipalDetails
- UserDetailsService 구현체 PrincipalDetailsService
- ApplicationConfig 으로 기존 bean 이동 및 등록
- Security Config 수정 (세션방식 x, jwt 필터 등록)
- jwt 인증 인가 필터 등록
- jwt 관련 서비스 구현
- 로그인 서비스 구현
- 컨트롤러 테스트 케이스 jwtService 의존성 이슈로 mockBean 등록


resolved #71

jinseong205 committed:
feat: security config 설정

- spring security 추가
- controller test case add filter= false

jinseong205 committed:
feat :  JWT 인증 인가 방식 추가

- UserDetails 구현체 PrincipalDetails
- UserDetailsService 구현체 PrincipalDetailsService
- ApplicationConfig 으로 기존 bean 이동 및 등록
- Security Config 수정 (세션방식 x, jwt 필터 등록)
- jwt 인증 인가 필터 등록
- jwt 관련 서비스 구현
- 로그인 서비스 구현
- 컨트롤러 테스트 케이스 jwtService 의존성 이슈로 mockBean 등록

jinseong205 committed:
principalDetails @Getter 추가

- principalDetails @Getter 추가

daengdaengLee commented on src/main/java/hello/until/jwt/JwtService.java:
```diff
@@ -0,0 +1,73 @@
+				.setSubject(userDetails.getUsername())
+				.setIssuedAt(new Date(System.currentTimeMillis()))
+				.setExpiration(new Date(System.currentTimeMillis() + jwtExpiration))	
+				.signWith(getSignInKey(), SignatureAlgorithm.HS256)
```
제 기억이 맞다면 알고리즘 명시 안 하고 전달한 키에 따라서 jjwt 가 알아서 판단하도록 하는 걸 추천했던 것 같아요.

jinseong205 replied:
해당 부분 수정하였습니다.

daengdaengLee commented on src/main/java/hello/until/user/dto/response/LoginResponse.java:
```diff
@@ -0,0 +1,16 @@
+@NoArgsConstructor
+public class LoginResponse {
+
+  @JsonProperty("access_token")
```
저희 json 에서도 camel case 쓰고 있지 않나요?

jinseong205 replied:
수정해놓도록 하겠습니다!

daengdaengLee commented on src/main/java/hello/until/user/service/UserService.java:
로그인을 filter 에서 처리하는 것도 봤던 것 같아요.
지금처럼 하면 user/service 패키지가 spring security 에 강결합하는 형태라서요...

jinseong205 replied:
이 부분은 분리해서 진행하도록 하겠습니다.

jinseong205 replied:
건호님 해당 부분 진행하려는데 제 기준 생각보다 시간이 좀 걸려서요 일단 
다른 resolve 먼저 해결하여 반영 후 수정해도 괜찮을까요?

daengdaengLee replied:
네네 괜찮습니다!

daengdaengLee commented on src/test/java/hello/until/item/controller/ItemControllerTest.java:
```diff
@@ -44,6 +58,9 @@ class ItemControllerTest {
     private Item testItem;
     private final DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS");
 
+    @MockBean
+    private JwtService jwtService;
```
요거 mock bean 추가한 건 어떻게 쓰이는걸까요?

jinseong205 replied:
이 부분이 없으면 컨트롤러 테스트가 케이스가 모두 실패해서
pr올리기 전에 추가한 부분입니다.
한번 더 확인해보겠습니다.

daengdaengLee replied:
아하,,, 이걸 명시해 줘야 mockbean 이 등록되나 보네요. 아마 내부에서 리턴값을 쓰는 데가 없어서 문제가 없었나봐요.
나중에 인증, 인가 내용 테스트 추가할 때 더 보면 될 것 같습니다!

daengdaengLee reviewed (dismissed):
고생하셨습니다!! LGTM!
코멘트에 질문만 확인 부탁드려요!

dudwns9280 commented on src/main/java/hello/until/user/service/UserService.java:
```diff
@@ -23,6 +28,22 @@ public class UserService {
+		authenticationManager
+				.authenticate(new UsernamePasswordAuthenticationToken(email, password));
+		
+		User user = userRepository.findByEmail(email).orElseThrow();
```
email로 user를 찾지 못했을 때 Exception이 필요할꺼 같아보여요!

jinseong205 replied:
해당내용 적용하도록 하겠습니다!

dudwns9280 commented on src/test/java/hello/until/user/service/UserServiceTest.java:
```diff
@@ -44,7 +50,7 @@ public class UserServiceTest {
 		
 	    passwordEncoder = new BCryptPasswordEncoder();
-		userService = new UserService(userRepository, passwordEncoder);
+		userService = new UserService(userRepository, passwordEncoder, jwtService, authenticationManager );
```
테스트 대상이 되는 bean은 @InjectMocks를 활용하면 bean 주입이 좀 더 깔끔할 거 같습니다!

jinseong205 replied:
해당 부분 수정해놓도록 하겠습니다!

jinseong205 committed:
리뷰 내용 수정

- jwt service 암호화 알고리즘 자동화
- LoginResponse 토큰 응답 변수 cammel case 변경
- usreService - login 예외처리 추가
- UserServiceTest mock 객체로 di
